### ISSUE TITLE
Istio 1.3+ on GKE requires additional firewall rules for validating webhooks to work

### ISSUE THREAD

**Issue by jroper** [istio/istio#19532]
In Istio 1.3, the `istio-galley` service was changed from using 443 as the target port, to using 9443 as the target port, while the service port is still 433. This causes Istio installs on GKE to run into this issue:

https://github.com/kubernetes/kubernetes/issues/79739

To summarise that issue, in GKE, the default firewall rules are that master nodes can only access nodes on ports 443 or 10250, as documented here:

https://cloud.google.com/kubernetes-engine/docs/how-to/private-clusters#add_firewall_rules

But the service port is 443 you say! Except that GKE enables aggregator routing:

https://github.com/kubernetes/kubernetes/blob/0d58709016f3358f34a52c12087d707a64cc6cb0/cluster/gce/gci/configure-kubeapiserver.sh#L95

Which means that the api server will route requests to the endpoint (port 9443), not the service (port 443) (see the above [issue](https://github.com/kubernetes/kubernetes/issues/79739) for a more detailed explanation of why etc).

The result of this is that when you try to create a `VirtualService` or `Gateway` or anything, you get a 504 gateway timeout, because the kube api server is blocked by the firewall from connecting to the validating webook.

So, basically, a default Istio 1.3+ install is broken on GKE. There are two workarounds:

* Disable validation (not a good work around, validation exists for a reason)
* Configure custom firewall rules to allow connection to the galley on port 9443

In either case, this should be documented in the instructions for how to install Istio on GKE. Ideally, Istio could solve this by using port 443 on the node, then the custom firewall rules (which are not straight forward to configure) are not needed.

**Comment by howardjohn** [comment]
I have never encountered this on GKE before. Any idea if there is some certain configuration of the cluster/project that requires the rule to be added?

**Comment by jroper** [comment]
Not sure, but my money is on this being the case for [vpc native clusters](https://cloud.google.com/kubernetes-engine/docs/how-to/alias-ips).

**Comment by jroper** [comment]
I'm wrong, it's not vpc native, it has to be a [private cluster](https://cloud.google.com/kubernetes-engine/docs/how-to/private-clusters), so nodes don't have external IP addresses. That page already mentions other projects that have this same issue, see https://cloud.google.com/kubernetes-engine/docs/how-to/private-clusters#add_firewall_rules.

**Comment by howardjohn** [comment]
Bringing discussion over from https://github.com/istio/istio/issues/20454. I have a method to reliably detect this: `kubectl get --raw /api/v1/namespaces/istio-system/services/https:istio-galley:https-validation/proxy/ready -v9 --request-timeout=2s`. The next step is finding out how to expose this (possibly under `istioctl analyze`)

**Comment by ayj** [comment]
This could be a good candidate for `istioctl precheck` (see https://docs.google.com/document/d/1x7eFR7e7yw_M1KB0ds5XUQPqai7HeSlnNEUyq8dpFeI/edit).

cc @esnible @irisdingbj

**Comment by howardjohn** [comment]
We need some target on the correct port (9443/15017) though, which
generally means after the install has completed. Any ideas to get around
that? I agree in principle though, if we can do it pre-install that is even
better

**Comment by ayj** [comment]
Good point. I think we'd need to install a test service/pod to verify pre-install. Another option would be to add this to a post-install check (e.g. verify-install).

**Comment by howardjohn** [comment]
Another -1 for `analyze`: `services/proxy` is not currently granted to istiod/galley. This is basically root access so we probably don't want to grant it

**Comment by howardjohn** [comment]
If we have a post-install check that is perfect

**Comment by jasonwzm** [comment]
Also the current scope of `istioctl analyze` is static configs.

**Comment by howardjohn** [comment]
verify-install looks like a pre-check, not a postcheck. there is also `post-install` for managing webhooks. Are we going to add a new postcheck or something? The doc linked was also about pre-checking

**Comment by haf** [comment]
@ayj Can you please make sure design docs are public, if you link to them as a foundation for future work here? :)

**Comment by ayj** [comment]
@haf - all of Istio's docs are in a shared Google drive. You can get access by joining the [istio-team-drive-access@](https://groups.google.com/forum/#!forum/istio-team-drive-access) group. Joining is instant and no approval is required.

**Comment by upodroid** [comment]
```
 Error creating: Internal error occurred: failed calling webhook "sidecar-injector.istio.io": Post https://istio-sidecar-injector.istio-system.svc:443/inject?timeout=30s: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
```

I spent a few hours scratching my head for why a vanilla install of Istio on Private GKE Cluster wasn't working. I learnt from earlier work with bank vaults to add 8443 if you have mutating admission webhooks to the fw rule for the master-to-nodes rule. Adding 9443 cleared the above error too.

**Comment by ayj** [comment]
cc @maasen

**Comment by AsherPemberton** [comment]
This firewall rule needs setting to port 15017 instead of 9443 on Istio 1.5

**Comment by upodroid** [comment]
I think we can close this out. https://istio.io/docs/setup/platform-setup/gke/ has instructions for amending the firewall for private gke clusters.

**Comment by taisph** [comment]
```
Error creating: Internal error occurred: failed calling webhook "sidecar-injector.istio.io": Post https://istio-sidecar-injector.istio-system.svc:443/inject?timeout=30s: context deadline exceeded
```

I got this issue after upgrading istio from 1.4 to 1.5.10 (using helm).

@upodroid said:
> [...] Adding 9443 cleared the above error too.

@APemberton1 said:
> This firewall rule needs setting to port 15017 instead of 9443 on Istio 1.5

Port 15017 was already allowed. I had to add port 9443 to the firewall instead contrary to the [Istio GKE docs](https://istio.io/latest/docs/setup/platform-setup/gke/). Does someone have an explanation as to why there seem to be confusion between these two ports?

**Comment by upodroid** [comment]
Did you read the right istio documentation? There are different docs for different versions.

The ports changed recently.

**Comment by taisph** [comment]
> Did you read the right istio documentation? There are different docs for different versions.
> The ports changed recently.

Not sure what you mean. There's no version on the docs I linked to. And the statement here was that in 1.5 this changed from 9443 to 15017 - which is the opposite of what I experienced.

**Comment by howardjohn** [comment]
>  https://istio-sidecar-injector.istio-system.svc:443/inject?timeout=30s
<https://istio-sidecar-injector.istio-system.svc/inject?timeout=30s>

This means you are not using Istiod. The doc is probably only correct for
Istiod, not legacy istio-sidecar-injector. Since 1.5 is EOL
https://istio.io/latest/news/support/announcing-1.5-eol-final/ the docs are
frozen though

**Comment by taisph** [comment]
> <https://istio-sidecar-injector.istio-system.svc/inject?timeout=30s> This means you are not using Istiod. The doc is probably only correct for Istiod, not legacy istio-sidecar-injector. Since 1.5 is EOL https://istio.io/latest/news/support/announcing-1.5-eol-final/ the docs are frozen though

I've not had much success switching to istiod which is why the cluster is still stuck on 1.4/1.5. I was just curious if something might be broken in the existing installation (ie. port configurations) which could explain the trouble we have switching to istiod.

Thank you.

**Comment by OneideLuizSchneider** [comment]
> ```
> Error creating: Internal error occurred: failed calling webhook "sidecar-injector.istio.io": Post https://istio-sidecar-injector.istio-system.svc:443/inject?timeout=30s: context deadline exceeded
> ```
> 
> I got this issue after upgrading istio from 1.4 to 1.5.10 (using helm).
> 
> @upodroid said:
> 
> > [...] Adding 9443 cleared the above error too.
> 
> @APemberton1 said:
> 
> > This firewall rule needs setting to port 15017 instead of 9443 on Istio 1.5
> 
> Port 15017 was already allowed. I had to add port 9443 to the firewall instead contrary to the [Istio GKE docs](https://istio.io/latest/docs/setup/platform-setup/gke/). Does someone have an explanation as to why there seem to be confusion between these two ports?

Me too, I had to add port 9443 to the firewall.
